### DISCUSSION TITLE
Problem with Android NDK build (BoringSSL or OpenSSL)

### DISCUSSION THREAD

**Discussion by ghost** [strongswan/strongswan#349] [Q&A] [answered]
I followed the tutorial here https://github.com/Ed-Yang/strongswan-android

Everything goes well, except for these error upon building the "armeabi-v7a" architecture.

if I use BoringSSL
.
.
.
[armeabi-v7a] SharedLibrary  : libstrongswan.so
ld: error: undefined symbol: EC_POINT_set_affine_coordinates
>>> referenced by openssl_ec_diffie_hellman.c:510 (../../../../../../../src/libstrongswan/plugins/openssl/openssl_ec_diffie_hellman.c:510)
>>>               /Users/admin/strongswan-android/strongswan/src/frontends/android/app/src/main/obj/local/armeabi-v7a/objs/strongswan/plugins/openssl/openssl_ec_diffie_hellman.o:(ec_group_new_brainpool)
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [/Users/admin/adt-bundle-mac-x86_64-20140702/sdk/ndk/22.1.7171670/build/core/build-binary.mk:728: /Users/admin/strongswan-android/strongswan/src/frontends/android/app/src/main/obj/local/armeabi-v7a/libstrongswan.so] Error 1

if I used OpenSSL
.
.
.
[armeabi-v7a] Compile thumb  : crypto_static <= aes-armv4.S
/Users/admin/strongswan-android/strongswan/src/frontends/android/app/src/main/jni/openssl/crypto/aes/asm/aes-armv4.S:358:2: error: out of range immediate fixup value
 sub r10,r3,#AES_encrypt-AES_Te @ Te
 ^
/Users/admin/android/strongswan/src/frontends/android/app/src/main/jni/openssl/crypto/aes/asm/aes-armv4.S:994:2: error: out of range immediate fixup value
 sub r10,r3,#AES_decrypt-AES_Td @ Td
 ^
make: *** [/Users/admin/adt-bundle-mac-x86_64-20140702/sdk/ndk/22.1.7171670/build/core/build-binary.mk:477: /Users/admin/strongswan-android/strongswan/src/frontends/android/app/src/main/obj/local/armeabi-v7a/objs/crypto_static/crypto/aes/asm/aes-armv4.o] Error 1

**Comment by deiger** [comment]
`EC_POINT_set_affine_coordinates` is defined in BoringSSL, including the Android repository, see [here](https://android.googlesource.com/platform/external/boringssl/+/refs/heads/master/src/include/openssl/ec.h#249) and [here](https://android.googlesource.com/platform/external/boringssl/+/refs/heads/master/src/crypto/fipsmodule/ec/ec.c#888).
Which version are you linking against?

**Reply by ghost** [reply]
Thank you!
Yes its defined there in your links. But I am referring to to the official documentation as it requires us to download the library from their git repo.
https://wiki.strongswan.org/projects/strongswan/wiki/AndroidVPNClientBuild

do check the BoringSSL git files and have a good look.
git clone git://git.strongswan.org/android-ndk-boringssl.git -b ndk-static openssl

only the EC_POINT_set_affine_coordinates_GFp is defined.

![Screen Shot 2021-04-26 at 7 47 00 PM](https://user-images.githubusercontent.com/34568349/116077575-37840100-a6c8-11eb-8bda-4a2c76cc8221.png)

**Comment by ghost** [comment]
for the OpenSSL library
git clone git://git.strongswan.org/android-ndk-openssl.git -b ndk-static openssl
according to this thread https://github.com/openssl/openssl/issues/7878 turning off thumb support should work but, now I get this new error

[armeabi-v7a] Compile arm    : crypto_static <= armv4-gf2m.S
/Users/admin/strongswan-android/strongswan/src/frontends/android/app/src/main/jni/openssl/crypto/bn/asm/armv4-gf2m.S :99:2: error: invalid instruction, any one of the following would fix this:
 vmov.32 d26, r2, r1
 ^
/Users/admin/strongswan-android/strongswan/src/frontends/android/app/src/main/jni/openssl/crypto/bn/asm/armv4-gf2m.S :99:15: note: invalid operand for instruction
 vmov.32 d26, r2, r1
              ^
/Users/admin/strongswan-android/strongswan/src/frontends/android/app/src/main/jni/openssl/crypto/bn/asm/armv4-gf2m.S :99:6: note: invalid operand for instruction
 vmov.32 d26, r2, r1
     ^
/Users/admin/strongswan-android/strongswan/src/frontends/android/app/src/main/jni/openssl/crypto/bn/asm/armv4-gf2m.S: 100:2: error: invalid instruction, any one of the following would fix this:
 vmov.32 d27, r12, r3
 ^
/Users/admin/strongswan-android/strongswan/src/frontends/android/app/src/main/jni/openssl/crypto/bn/asm/armv4-gf2m.S: 100:15: note: invalid operand for instruction
 vmov.32 d27, r12, r3
              ^
/Users/admin/strongswan-android/strongswan/src/frontends/android/app/src/main/jni/openssl/crypto/bn/asm/armv4-gf2m.S: 100:6: note: invalid operand for instruction
 vmov.32 d27, r12, r3
     ^
make: *** [/Users/admin/adt-bundle-mac-x86_64-20140702/sdk/ndk/22.1.7171670/build/core/build-binary.mk:477: /Users/admin/strongswan-android/strongswan/src/frontends/android/app/src/main/obj/local/armeabi-v7a/objs/crypto_static/crypto/bn/asm/armv4-gf2m.o] Error 1

**Reply by tobiasbrunner** [reply]
Don't use that repository. It's even older than the BoringSSL fork. Try the latter again with the current master as there was an incorrect version check related to `EC_POINT_set_affine_coordinates*`.

**Comment by ghost** [comment, marked as answer]
> Don't use that repository. It's even older than the BoringSSL fork. Try the latter again with the current master as there was an incorrect version check related to `EC_POINT_set_affine_coordinates*`.

Resolved. Thanks!
